src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc

```diff
@@ -0,0 +1,190 @@
+#include <string>
+#include <vector>
+
+#include "../../../../runtime/file_utils.h"
+#include "../../../../target/source/codegen_c.h"
+
+namespace tvm {
+namespace runtime {
+
+using namespace tir;
+
+class CodeGenCMSISNN : public tvm::codegen::CodeGenC {
+ public:
+  void Init(bool output_ssa) {
+    decl_stream << "#include <stdio.h>\n";
+    decl_stream << "#include <stdlib.h>\n";
+    decl_stream << "#include <dlpack/dlpack.h>\n";
+    decl_stream << "#include <tvm/runtime/crt/module.h>\n";
+    decl_stream << "#include <arm_nnfunctions.h>\n";
+    CodeGenC::Init(output_ssa);
+  }
+
+  /*!
+   * \brief Emit code that offloads a subgraph to the Cortex-M
+   *
+   * \return string of code that offloads a subgraph to the Cortex-M
+   */
+  void AddFunction(const PrimFunc& prim_func) {
+    PrintExternCPrefix(stream);
+    CodeGenC::AddFunction(prim_func);
+    PrintExternCPostfix(stream);
+  }
+
+ private:
+  void VisitExpr_(const CallNode* op, std::ostream& os) {  // NOLINT(*)
+    if (!op->op.same_as(builtin::call_extern())) {
+      return;
+    }
+    std::string cmsis_func_name = op->args[0].as<StringImmNode>()->value;
+    if (cmsis_func_name == "arm_softmax_s8") {
+      EmitSoftmax(op);
+    }
+    return;
+  }
+
+  /*!  * \brief Creates a cplusplus guard prefix for extern "C" printing */
+  void PrintExternCPrefix(std::ostringstream& ss) {
+    PrintIndent();
+    ss << "#ifdef __cplusplus\n";
+    ss << "extern \"C\" {\n";
+    ss << "#endif\n";
+  }
+
+  /*!  * \brief Creates a cplusplus guard postfix for extern "C" printing */
+  void PrintExternCPostfix(std::ostringstream& ss) {
+    PrintIndent();
+    ss << "#ifdef __cplusplus\n";
+    ss << "}\n";
+    ss << "#endif\n";
+  }
+
+  /*!  * \brief Emits CMSIS-NN code block for softmax */
+  void EmitSoftmax(const CallNode* op) {
+    // @tir.call_extern("arm_softmax_s8", buffer_0, num_rows, row_size,
+    //                  mult, shift, diff_min, buffer_1, dtype=int8)
+    std::string cmsis_func_name = op->args[0].as<StringImmNode>()->value;
+    int32_t num_rows = op->args[2].as<IntImmNode>()->value;
+    int32_t row_size = op->args[3].as<IntImmNode>()->value;
+    int32_t mult = op->args[4].as<IntImmNode>()->value;
+    int32_t shift = op->args[5].as<IntImmNode>()->value;
+    int32_t diff_min = op->args[6].as<IntImmNode>()->value;
+
+    PrintIndent();
+    stream << "int32_t num_rows = " << num_rows << ";\n";
+    PrintIndent();
+    stream << "int32_t row_size = " << row_size << ";\n";
+    PrintIndent();
+    stream << "int32_t mult = " << mult << ";\n";
+    PrintIndent();
+    stream << "int32_t shift = " << shift << ";\n";
+    PrintIndent();
+    stream << "int32_t diff_min = " << diff_min << ";\n";
+    PrintIndent();
+    stream << cmsis_func_name << "(input,";
+    PrintIndent();
+    stream << " num_rows, row_size, mult, shift, diff_min, output);\n";
+    PrintIndent();
+    stream << "return;\n";
+  }
+};
+
+class CMSISNNModuleNode : public runtime::ModuleNode {
```

Comment:
Shall we break this into source_module.cc in cmsis to indicate runtime.Module that is a c-source ?

Reply:
As discussed, not needed in cmsisnn world.